src/core/watchQueryOptions.ts

```diff
@@ -164,31 +164,98 @@ export interface FetchMoreQueryOptions<TVariables, TData = any> {
   context?: DefaultContext;
 }
 
-export type UpdateQueryFn<
+export interface UpdateQueryFn<
+  TData,
+  TVariables extends OperationVariables,
+  TOptions = {},
+> {
+  (mapFn: UpdateQueryMapFn<TData, TVariables, TOptions>): void;
+}
+
+export type UpdateQueryOptions<TData, TVariables> = {
+  variables?: TVariables;
+} & (
+  | {
+      /**
+       * Indicate if the previous query result has been found fully in the cache.
+       */
+      complete: true;
+      previousQueryResult: Unmasked<TData>;
+    }
+  | {
+      /**
+       * Indicate if the previous query result has not been found fully in the cache.
+       * Might have partial or missing data.
+       */
+      complete: false;
+      previousQueryResult: DeepPartial<Unmasked<TData>> | undefined;
+    }
+);
+
+export interface UpdateQueryMapFn<
+  TData = any,
+  TVariables extends OperationVariables = OperationVariables,
+  TOptions = {},
+> {
+  (
+    /**
+     * @deprecated Use `options.previousQueryResult` instead.
+     */
```

Comment:
Can we be a bit more verbose here? I think having a bit of a description on why this is an issue will help users understand why these need to stop using this argument. Feel free to workshop this suggestion:

```suggestion
    /**
     * @deprecated This value is not type-safe and may contain partial data. This
     * argument will be removed in the next major version of Apollo Client. Use
     * `options.previousData` instead for a more type-safe value.
     */
```